Hash modulo meta

roberth opened the pull request:
<!--

IMPORTANT

Nix is a non-trivial project, so for your contribution to be successful,
it really is important to follow the contributing guidelines:

https://github.com/NixOS/nix/blob/master/CONTRIBUTING.md

Even if you've contributed to open source before, take a moment to read it,
so you understand the process and the expectations.

- what information to include in commit messages
- proper attribution
- volunteering contributions effectively
- how to get help and our review process.

PR stuck in review? We have two Nix team meetings per week online that are open for everyone in a jitsi conference:

- https://calendar.google.com/calendar/u/0/embed?src=b9o52fobqjak8oq8lfkhg3t0qg@group.calendar.google.com

-->

## Motivation

`meta` is lost at the derivation level, but it doesn't need to be that way, thanks to quotient hashing aka hash modulo.

This change makes use of the existing "discrepancy" between derivation hashes and output hashes to allow more information to be stored in derivations without causing rebuilds.

The core of the change is fairly simple; see the `hashDerivationModulo` hunk.

## Changes

- Add system feature `derivation-meta` improving hash quotient logic to support storing `meta`
- Adjust JSON formats to make `meta` a first class top level attr instead of the ATerm embedding

## Notes

- Must be opted in to by derivations/expressions, as usual, like ca derivations, structuredAttrs
- Not supported in non-structuredAttrs. Didn't seem worth the added complexity. Users should move to structuredAttrs instead and afaik Nixpkgs is on track to do so.

## Example

```console
$ nix run github:roberth/nix/hash-modulo-meta -- \
  build --store ~/stores/tmp \
    --extra-experimental-features derivation-meta \
    --impure --expr \
    'with import (builtins.fetchTarball { url = "https://github.com/roberth/nixpkgs/archive/derivation-meta.tar.gz"; }) { config.derivationMeta = true; }; python3'
...
<progress bar, substituting from cache.nixos.org>
...
```

```console
$ nix run github:roberth/nix/hash-modulo-meta -- derivation show --store ~/stores/tmp --extra-experimental-features derivation-meta --impure --expr 'with import (builtins.fetchTarball { url = "https://github.com/roberth/nixpkgs/archive/derivation-meta.tar.gz"; }) { config.derivationMeta = true; }; python3'
{
  "vcllzyc6d87wxlncjcwxlgy82lna4h4v-python3-3.13.9.drv": {
    "args": ...
    ...
    "meta": {
      "available": true,
      "broken": false,
      "changelog": "https://docs.python.org/release/3.13.9/whatsnew/changelog.html",
...
  }
}
```


## Context

- Step towards correctly implementing `meta.timeout`
- Implements, closes https://github.com/NixOS/nix/issues/10780
- Incremental change on topic of https://github.com/NixOS/nix/issues/9774
- Closes perhaps https://github.com/NixOS/nix/issues/13564
- Possible solution to #4677
<!-- Provide context. Reference open issues if available. -->

<!-- Non-trivial change: Briefly outline the implementation strategy. -->

<!-- Invasive change: Discuss alternative designs or approaches you considered. -->

<!-- Large change: Provide instructions to reviewers how to read the diff. -->

---

Add :+1: to [pull requests you find important](https://github.com/NixOS/nix/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc).

The Nix maintainer team uses a [GitHub project board](https://github.com/orgs/NixOS/projects/19) to [schedule and track reviews](https://github.com/NixOS/nix/tree/master/maintainers#project-board-protocol).

roberth committed:
Add derivation-meta feature for no-rebuild meta stored in derivation

roberth committed:
Translate between __meta and meta in JSON derivation

roberth committed:
Only activate derivation-meta when __meta is an object

roberth committed:
Update derivation-options-v1 schema for derivation-meta

roberth committed:
Ensure backward compatibility for meta field in DerivationOptions JSON

Make the meta field optional in DerivationOptions JSON deserialization to maintain backward compatibility with JSON that predates the derivation-meta feature.

Changes:
- Use optionalValueAt instead of valueAt for meta field deserialization
- Mark meta as optional in derivation-options-v1 schema
- Add unit test for backward compatibility

roberth committed:
Extract __meta to top-level meta field in Derivation JSON format

When the derivation-meta experimental feature is enabled, extract the
__meta field from structuredAttrs to a top-level meta field in the
Derivation JSON format (used by `nix derivation show/add`). This makes
the format consistent with DerivationOptions JSON and provides better
visibility of metadata without it being buried in structuredAttrs.

The implementation:
- Creates derivationToJson() helper accepting ExperimentalFeatureSettings
- Extracts __meta to top-level meta field when feature is enabled
- Requires experimental feature when deserializing JSON with meta field
- Reconstructs __meta in structuredAttrs during deserialization
- Maintains backward compatibility with JSON lacking meta field

Tests verify both serialization and deserialization with mocked
experimental feature settings, and functional tests ensure the
experimental feature requirement is enforced.

roberth committed:
Add functional test that __meta doesn't make it into the sandbox

roberth committed:
doc: Apply suggestions

Thank you fricklerhandwerk!

Co-authored-by: Valentin Gagarin <valentin@gagarin.work>

Ericson2314 commented on src/libstore/build/derivation-goal.cc:
```diff
@@ -73,6 +73,23 @@ Goal::Co DerivationGoal::haveDerivation(bool storeDerivation)
+            experimentalFeatureSettings.require(Xp::DerivationMeta);
+        }
+    }
+
```
I would not put this here but instead make it part of the derivation options parsing logic. Then we will fail harder, which is good.

Ericson2314 commented on src/libstore/derivation-options.cc:
```diff
@@ -343,6 +343,18 @@ DerivationOptions<SingleDerivedPath> derivationOptionsFromStructuredAttrs(
+            // Only extract __meta if derivation-meta feature is used
+            if (hasDerivationMetaFeature(structuredAttrs)) {
+                return getObject(structuredAttrs.at("__meta"));
+            }
```
The error checking for `__meta` with the feature would go here.

Ericson2314 commented on src/libstore/derivations.cc:
```diff
@@ -1552,44 +1610,36 @@ static void basicDerivationToJson(json & res, const BasicDerivation & d)
+            if (nix::hasDerivationMetaFeature(structuredAttrs)) {
+                auto metaIt = structuredAttrs.find("__meta");
+                res["meta"] = metaIt->second;
+                structuredAttrs.erase(metaIt);
```
Hmm this is going to be a bit janky

Ericson2314 replied:
this crates a "have we parsed the options yet" race condition / state of sorts. We are better off eagerly extracting it at ATerm parse time as we do with structured attrs.

roberth committed:
Extract __meta at parse time

It used to defer this to consumers. By performing the transformation
immediately upon ingestion we avoid some confusion about the exact
meaning of e.g. structuredAttrs at potentially many points in the code.
We can now trust that it only has the "true" structured attributes that
go into the builder.

This may make ill-formed derivations harder to handle, but that is a
price worth paying for added clarity going forward.

Now, invalid __meta usage (missing experimental feature, unsorted
requiredSystemFeatures) is caught immediately when the derivation is
parsed or instantiated.

To make re-injection into requiredSystemFeatures deterministic, it must
be sorted when derivation-meta is used in instantiation. This avoids
storing the original array position, and removes unnecessary entropy
from derivations.

Meta moves from DerivationOptions to BasicDerivation, since it is a
derivation-level concept, not a build option.

roberth committed:
Improve error messages when derivation-meta is unsupported

Old daemons and remote builders don't understand derivation-meta,
and would produce hash mismatch errors when receiving and validating
such derivations.

A new worker protocol feature marks support for ingesting these
new derivations.

The client checks for this feature in writeDerivation (for daemon)
and build-remote (for remote builders) before a hash mismatch
would occur, so that it can report a more specific error message.

roberth committed:
tests: filter 'waiting for' messages in nix-profile test

The 'waiting for another Nix process to finish fetching input' message
appears nondeterministically, causing the diff comparison to fail.

roberth commented:
- Implemented the review. Thank you @Ericson2314 

- I've employed worker protocol features to improve the error message when the daemon is old. (was hash mismatch in output)

- Had to include #15848 for the moment.

edolstra commented:
FWIW, Determinate Nix has a `derivationWithMeta` builtin: https://github.com/DeterminateSystems/nix-src/pull/357

This is currently used in Determinate Secure Packages to add meta attributes to the provenance of derivations (i.e. `stdenv.mkDerivation` calls `derivationWithMeta` with the subset of the meta attributes that we want to be recorded in the Nix database / binary cache).

roberth commented:
I like that this doesn't require extra bookkeeping.
It's hardly a tradeoff in my perspective. The added hashing complexity is minor and hardly exposed to users (in fact beneficial to blast away the 1:1 deriver misconception). On top of that you get a simpler data model for users: just derivations.
Actually the deriver problem goes deeper. Using this method, there's no way for the metadata to become ambiguous because of old, stupid or evil derivations that may instantiate before or after.

Ericson2314 commented on src/libstore/build/derivation-goal.cc:
```diff
@@ -62,6 +62,15 @@ Goal::Co DerivationGoal::haveDerivation(bool storeDerivation)
+            "derivation '%s' has '__meta' attribute but does not require 'derivation-meta' system feature",
+            worker.store.printStorePath(drvPath));
+    }
+
```
This is the one I was hoping to move to derivation options?

roberth replied:
Oops!

Done: https://github.com/NixOS/nix/compare/25286c3cf28529ac679f79c3d74ac3498bfd3441..135bc5e827f92dab44eaf9b1e913abda108b32d4

roberth committed:
JSON_IMPL_WITH_XP_FEATURES: add xpSettings to to_json

This makes to_json symmetric with from_json, which already accepts
ExperimentalFeatureSettings. Types whose serialization depends on
experimental features can now receive the settings in to_json as well,
instead of having to use a separate free function.

All existing implementations pass the parameter through unused for now.

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nix-derivation-0-1/79491/2

domenkozar commented:
What is blocking this? I'm happy to help.

roberth commented:
I've done the big rebase. Fresh PR felt right to me: https://github.com/NixOS/nix/pull/16285

Haven't got around to doing the practical testing again.
@domenkozar, want to do a user test of #16285?

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/dynamic-derivation-working-group-meeting-notes-1/79575/1
